salt/modules/file.py

```diff
@@ -1881,132 +1911,107 @@ def line(path, content=None, match=None, mode=None, location=None,
         match = content
 
     with salt.utils.files.fopen(path, mode='r') as fp_:
-        body = salt.utils.stringutils.to_unicode(fp_.read())
-    body_before = hashlib.sha256(salt.utils.stringutils.to_bytes(body)).hexdigest()
+        body = [salt.utils.stringutils.to_unicode(line) for line in fp_.readlines()]
+    body_before = hashlib.sha256(salt.utils.stringutils.to_bytes(''.join(body))).hexdigest()
+    if body and _get_eol(body[-1]):
+        body.append('')
+
     after = _regex_to_static(body, after)
     before = _regex_to_static(body, before)
     match = _regex_to_static(body, match)
 
     if os.stat(path).st_size == 0 and mode in ('delete', 'replace'):
         log.warning('Cannot find text to {0}. File \'{1}\' is empty.'.format(mode, path))
-        body = ''
+        body = []
     elif mode == 'delete':
-        body = os.linesep.join([line for line in body.split(os.linesep) if line.find(match) < 0])
+        body = [line for line in body if line != match[0]]
```

Comment:
If there are no matching lines, `match` will be an empty string and this will trigger an `IndexError`. I also do not understand why you're comparing this only to `match[0]`. If there are multiple matches, don't we want them all removed?

Reply:
Fixed.